src/multiversioning.zig

```diff
@@ -911,16 +940,46 @@ pub const Multiversion = struct {
         result: IO.ReadError!usize,
     ) void {
         assert(self.stage == .source_read);
-
-        posix.close(self.source_fd.?);
-        self.source_fd = null;
+        assert(self.source_fd != null);
+        assert(self.source_offset != null);
+        assert(self.source_size != null);
+        assert(self.source_offset.? < self.source_size.?);
+
+        defer {
+            if (self.stage != .source_read) {
+                assert(self.stage == .err or self.stage == .ready);
+                assert(self.source_fd != null);
+                assert(self.source_offset != null);
+                assert(self.source_size != null);
+
+                posix.close(self.source_fd.?);
+                self.source_offset = null;
+                self.source_size = null;
+                self.source_fd = null;
+            }
+        }
 
         const bytes_read = result catch |e| return self.handle_error(e);
-        const source_buffer = self.source_buffer[0..bytes_read];
-
-        self.stage = .target_update;
-        self.target_update(source_buffer) catch |e| return self.handle_error(e);
-        assert(self.stage == .ready);
+        self.source_offset.? += bytes_read;
+        assert(self.source_offset.? <= self.source_buffer.len);
+
+        // The file is smaller than we expected.
+        // It was likely modified in the time since our statx().
+        if (bytes_read == 0) return self.handle_error(error.FileModifiedDuringRead);
+
+        switch (std.math.order(self.source_offset.?, self.source_size.?)) {
```

Comment:
An alternative here is not to bother with `statx.size`, and just read until read returns `0`. But, given we _have_ `statx`, we might as well use it.

Reply:
Ah, actually I'll go with your suggestion -- we only have statx on Linux.